File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.py
from collections import deque

def solution(maps):
    
    dx = (0, 0, -1, 1)
    dy = (-1, 1, 0, 0)
    
    N = len(maps); M = len(maps[0])
    
    def bfs(sx, sy, ex, ey):
        queue = deque([]); queue.append((sx, sy))
        visit = [[-1 for m in range(M)] for n in range(N)]; visit[sx][sy] += 1
        
        while queue:
            x, y = queue.popleft()
            for i in range(4):
                nx = x + dx[i]; ny = y + dy[i]
                if nx < 0 or nx >= N or ny < 0 or ny >= M: continue
                if maps[nx][ny] == 'X' or visit[nx][ny] != -1: continue
                visit[nx][ny] = visit[x][y] + 1
                queue.append((nx, ny))
        return visit[ex][ey]
            
            
    start = tuple(); end = tuple(); lever = tuple()
    for i in range(N):
        for j in range(M):
            if maps[i][j] == 'S': start = (i, j)
            elif maps[i][j] == 'E': end = (i, j)
            elif maps[i][j] == 'L': lever = (i, j)
                
    pre_count = bfs(start[0], start[1], lever[0], lever[1])
    if pre_count == -1: return -1
    post_count = bfs(lever[0], lever[1], end[0], end[1])
    if post_count == -1: return -1
    return pre_count + post_count

File: 프로그래머스/unrated/159993. 미로 탈출/미로 탈출.cpp
#include <string>
#include <vector>
#include <iostream>
#include <limits.h>
#include <queue>
#define MAX 100
using namespace std;

int N, M;
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

int bfs(vector<string> &maps, pair<int, int> start, pair<int, int> end){
    queue<pair<pair<int, int>, int>> queue;
    queue.push({{start.first, start.second}, 0});
    
    int visit[MAX][MAX] = {false,};
    visit[start.first][start.second] = true;
    
    while (!queue.empty()){
        int x = queue.front().first.first;
        int y = queue.front().first.second;
        int depth = queue.front().second;
        queue.pop();
        
        if (x == end.first && y == end.second){
            return depth;
        }
        
        for (int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (0 <= nx && nx < N && 0 <= ny && ny < M){
                if (maps[nx][ny] != 'X' && visit[nx][ny] == false){
                    queue.push({{nx, ny}, depth+1});
                    visit[nx][ny] = true;
                }
            }
        }
    }
    return -1;
}

int solution(vector<string> maps) {
    N = maps.size();
    M = maps[0].length();
    
    pair<int, int> S, E, L;
    for (int i=0; i<N; i++){
        for (int j=0; j<M; j++){
            if (maps[i][j] == 'S'){
                S = {i, j};
            }
            else if (maps[i][j] == 'E'){
                E = {i, j};
            }
            else if (maps[i][j] == 'L'){
                L = {i, j};
            }
        }
    }
    
    int pre_count = bfs(maps, S, L);
    if (pre_count == -1) return -1;
    int post_count = bfs(maps, L, E);
    if (post_count == -1) return -1;
    return pre_count+post_count;
}